amethyst_assets/src/storage.rs

```diff
@@ -0,0 +1,230 @@
+        if self.bitset.contains(handle.id()) {
+            Some(unsafe { self.assets.get_mut(handle.id()) })
+        } else {
+            None
+        }
+    }
+
+    /// Process finished asset data and maintain the storage.
+    pub fn process<F>(&mut self, mut f: F, errors: &Errors)
+    where
+        F: FnMut(A::Data) -> Result<A, BoxedErr>,
+    {
+        while let Some(processed) = self.processed.try_pop() {
+            let Processed {
+                data,
+                format,
+                handle,
+                name,
+            } = processed;
+            let assets = &mut self.assets;
+            let bitset = &mut self.bitset;
+            let handles = &mut self.handles;
+            errors.execute::<AssetError, _>(|| {
+                println!("Got asset with name {}", &name);
+
+                let asset = data.and_then(|d| f(d))
+                    .map_err(|e| AssetError::new(name, format, e))?;
+
+                let id = handle.id();
+                bitset.add(id);
+                handles.push(handle);
+
+                // NOTE: the loader has to ensure that a handle will be used
+                // together with a `Data` only once.
+                unsafe {
+                    assets.insert(id, asset);
+                }
+
+                Ok(())
+            });
+        }
+
+        while let Some(i) = self.handles.iter().position(Handle::is_unused) {
+            let old = self.handles.swap_remove(i);
+            let id = i as u32;
+            unsafe {
+                self.assets.remove(id);
+            }
+            self.bitset.remove(id);
+            self.unused_handles.push(old);
+
+            println!("Removed value!");
+        }
+    }
+}
+
+impl<A: Asset> Drop for AssetStorage<A> {
+    fn drop(&mut self) {
+        let bitset = &self.bitset;
+        unsafe { self.assets.clean(|id| bitset.contains(id)) }
+    }
+}
+
+/// A default implementation for an asset processing system
+/// which converts data to assets and maintains the asset storage
+/// for `A`.
+///
+/// This system can only be used if the asset data implements
+/// `Into<Result<A, BoxedErr>>`.
+pub struct Processor<A> {
+    marker: PhantomData<A>,
+}
+
+impl<A> Processor<A> {
+    /// Creates a new asset processor for
+    /// assets of type `A`.
+    pub fn new() -> Self {
+        Processor {
+            marker: PhantomData,
+        }
+    }
+}
+
+impl<'a, A> System<'a> for Processor<A>
+where
+    A: Asset,
+    A::Data: Into<Result<A, BoxedErr>>,
+{
+    type SystemData = (FetchMut<'a, AssetStorage<A>>, Fetch<'a, Errors>);
+
+    fn run(&mut self, (mut storage, errors): Self::SystemData) {
+        storage.process(Into::into, &errors);
+    }
+}
+
+/// A handle to an asset. This is usually what the
+/// user deals with, the actual asset (`A`) is stored
+/// in an `AssetStorage`.
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Hash(bound = ""))]
```

Comment:
I don't think use of the derivative crate is necessary here, I think we can just derive these.

Reply:
It is necessary, we also want to `Clone` handles which have an `A`sset type that's `!Clone`.